[feature] Play module initial implementation

ValentinMassa opened the pull request:
- Ive added a new field into Player struct
- Ive added two files(Stack and Play) with Headers and .c but they arent developed

ValentinMassa committed:
Changes:
- Ive added a new field into Player struct
-Ive added two files(Stack and Play) with Headers and .c but they arent developed

ValentinMassa reviewed (commented):
Aca deje el tema de play, analicenlo

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+    void *info;
+    unsigned sizeInfo;
+    struct sNode *next;
+} tNode;
```
`tNode`, and `sNode` structs must be named `Node`. But, Why? They are the same struct and is a convention to named classes using [CamelCase style](https://en.wikipedia.org/wiki/Camel_case). `C` doesn't have classes, but in the way that this `Node` works, it is a low level class with methods and attributes.

Also, use `__` to auto explain that those atributes are private, its an usual convention in `Python` and `JavaScript` (Google style) to mark those atributes as private in languages without a `private` reserved word.

It should be as follow:

```c
typedef struct Node {
    void* __data;
    size_t __sizeOfData;
    struct Node* __next;
} Node;
```

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+
+typedef tNode *tStack;
+
+void createStack(tStack *ps);  // ps is equal to pointer to stack
```
Avoid `create` instead use `new` as prefix.

hozlucas28 replied:
Also, `ps`? A better parameter name could be `stack`, using a prefix `p` is redundant, you can avoid the comment if do that. Remember that `C` is no case sensitive.

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+typedef tNode *tStack;
+
+void createStack(tStack *ps);  // ps is equal to pointer to stack
+int insertIntoStack(tStack *ps, const void *d, unsigned quantityOfBytes);
```
To much to comment... just check out this approach:
```c
size_t pushStack(Stack* stack, void* data, const size_t sizeOfData);
```

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+
+void createStack(tStack *ps);  // ps is equal to pointer to stack
+int insertIntoStack(tStack *ps, const void *d, unsigned quantityOfBytes);
+int grabElementFromStack(tStack *ps, void *d, unsigned quantityOfBytes);
```
Use `pop` as prefix (it's a convention in several languages) and parameters aren't auto explained. A correct way to write this function is:
```c
size_t popStack(Stack* stack, void* store, const size_t sizeOfStore);
```

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+void createStack(tStack *ps);  // ps is equal to pointer to stack
+int insertIntoStack(tStack *ps, const void *d, unsigned quantityOfBytes);
+int grabElementFromStack(tStack *ps, void *d, unsigned quantityOfBytes);
+int isTheStackEmpty(const tStack *ps);
```
`the` must be out, it doesn't add any value to the function name. Also, `int` as a return doesn't fit with a `true` or `false` return value. Better way:
```c
unsigned char isStackEmpty(const Stack* stack);
```

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+int insertIntoStack(tStack *ps, const void *d, unsigned quantityOfBytes);
+int grabElementFromStack(tStack *ps, void *d, unsigned quantityOfBytes);
+int isTheStackEmpty(const tStack *ps);
+int isTheStackFull(const tStack *ps, unsigned quantityOfBytes);
```
Same comments as the above. A better way could be:
```c
size_t isStackFull(const Stack* stack, const size_t sizeOfData);
```

ValentinMassa replied:
i think that u dont understand anything about c. cheers

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+int grabElementFromStack(tStack *ps, void *d, unsigned quantityOfBytes);
+int isTheStackEmpty(const tStack *ps);
+int isTheStackFull(const tStack *ps, unsigned quantityOfBytes);
+int ViewLastElementOfStack(const tStack *ps, void *d, unsigned quantityOfBytes);
```
Inconsistent function name, and `View` prefix is to general. Correct way...
```c
size_t getLastElementStack(Stack* stack, void* store, const size_t sizeOfStore);
```

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+int isTheStackEmpty(const tStack *ps);
+int isTheStackFull(const tStack *ps, unsigned quantityOfBytes);
+int ViewLastElementOfStack(const tStack *ps, void *d, unsigned quantityOfBytes);
+void destroyStack(tStack *ps);
```
```c
void destroyStack(Stack* stack);
```

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+    struct sNode *next;
+} tNode;
+
+typedef tNode *tStack;
```
Same convention as the comment above. `tStack` must be named `Stack`.

hozlucas28 commented on libs/Stack/stack.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef STACK_H_INCLUDED
+#define STACK_H_INCLUDED
```
Invalid header tag, it must be `LIBS__STACK_H_INCLUDED` to prevent tag collisions.

hozlucas28 commented on src/main.c:
```diff
@@ -6,6 +6,7 @@
 
 #include "./configuration/main.h"
 #include "./macros.h"
+#include "./play.h"
```
Invalid path, the correct one is `./play/main.h`. `main.h` because it's a module, if you use `play.h` inside `play` module is redundant, and we are using `main.h` as a [barrel file](https://dev.to/tassiofront/barrel-files-and-why-you-should-stop-using-them-now-bc4) to get all primitives from an inner module.

> In `C` there is no tree shaking by default.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+#include "play.h"
```
Incorrect path, use a safety path to import from the main root as the current file. Correct one: `./play.h`.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+#include "play.h"
+
+#include "../../libs/Stack/stack.h"
```
Violation of libs convention. You must include only the `main.h` file inside libs root, so correct route is `../../libs/main.h`.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+#include "../../libs/Stack/stack.h"
+#include "../structs.h"
+
+int playTicTacToe(const int gamesPerPlayer, char* nameOfTheLocalFile) {
```
Just receive the `config` variable inside [main.c](https://github.com/hozlucas28/C-Algorithms-Practical-Work-2025/blob/Master/src/main.c) as a parameter, because you are falling into an argument drilling. Correct way:
```c
unsigned char playTicTacToe(const Configuration* config)
```

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+#include "../structs.h"
+
+int playTicTacToe(const int gamesPerPlayer, char* nameOfTheLocalFile) {
+    tStack _ps;
```
Use an auto explained variable name... `players`. Also, it must be a `List` not a `Stack`, Why? because we need to get elements inside an specific index, and a `Stack` just pop his last element.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+
+int playTicTacToe(const int gamesPerPlayer, char* nameOfTheLocalFile) {
+    tStack _ps;
+    Player _player;
```
Unnecessary underscore, avoid it.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+int playTicTacToe(const int gamesPerPlayer, char* nameOfTheLocalFile) {
+    tStack _ps;
+    Player _player;
+    int games = 0;
```
Decrement games instead of starts it with `0`, and use `size_t`. Games mustn't have negative values, it's like declare an age that could be `-24`. Also, use the config argument in this way...

`size_t games = config->gamesPerPlayer;`

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+
+    createStack(&_ps);
+
+    askNamesAndMixPlayers(&_ps);  // this function will return the stack with the players mixed
```
Why did you explain what will happen inside `askNamesAndMixPlayers` function? It's unnecessary, today we mixed them, but tomorrow? we don't know. Also, it's a violation of [Single Responsibility](https://en.wikipedia.org/wiki/Single-responsibility_principle). Clean way...
```c
Queue playersQueue;

requestPlayers(&players); // Push to `players` list.

newQueue(&playersQueue);

randomSort(&players, &playersQueue); // Push `players` from list to `playersQueue` queue in a random way.
```

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+    askNamesAndMixPlayers(&_ps);  // this function will return the stack with the players mixed
+
+    while (grabElementFromStack(&_ps, &_player, sizeof(_player)))
+    // this function return 1 if its all ok, and 0 if the stack is empty
```
As the comment of `_players` variable... we only need to ask if `playersQueue` queue is empty.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+    while (grabElementFromStack(&_ps, &_player, sizeof(_player)))
+    // this function return 1 if its all ok, and 0 if the stack is empty
+    {
+        printf("Hello, Player %s, the game is about to star\n", _player.name);
```
Typo error, and invalid consistence with [main.c](https://github.com/hozlucas28/C-Algorithms-Practical-Work-2025/blob/Master/src/main.c) outputs.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+        printf("Hello, Player %s, the game is about to star\n", _player.name);
+        games = gamesPerPlayer;
+        while (games > 0) {
+            askPlayerIfHeIsReady(&_player);  // this function will interrogate
```
Avoid `games` initialization, and the function has a verbose name in english, named it `requestPlayerConfirmation`.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+                                             // the player if he is ready
+            playGame(&_player);
+
+            games--;
```
Avoid, info on configuration comment.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+        while (games > 0) {
+            askPlayerIfHeIsReady(&_player);  // this function will interrogate
+                                             // the player if he is ready
+            playGame(&_player);
```
Another function to play inside `play`? That's sound weird. Comment this line, in the next daily I will explain it.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+            playGame(&_player);
+
+            games--;
+            printf("\n %s, \n Your actual points are: %d \n", _player.name, _player.points);
```
Invalid consistence with [main.c](https://github.com/hozlucas28/C-Algorithms-Practical-Work-2025/blob/Master/src/main.c) outputs. Also, use separated `printf`, if you want to print more than one line to improve [DX](https://microsoft.github.io/code-with-engineering-playbook/developer-experience/).

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+
+            games--;
+            printf("\n %s, \n Your actual points are: %d \n", _player.name, _player.points);
+            printf("Next game, %d: games left\n", games);
```
Invalid consistence with [main.c](https://github.com/hozlucas28/C-Algorithms-Practical-Work-2025/blob/Master/src/main.c) outputs.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+            printf("Next game, %d: games left\n", games);
+        }
+
+        puts("You have finished your games");
```
Invalid consistence with [main.c](https://github.com/hozlucas28/C-Algorithms-Practical-Work-2025/blob/Master/src/main.c) outputs.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        puts("You have finished your games");
+        printf("%s, Your final points are: %d \n", _player.name, _player.points);
```
Invalid consistence with [main.c](https://github.com/hozlucas28/C-Algorithms-Practical-Work-2025/blob/Master/src/main.c) outputs.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+        puts("You have finished your games");
+        printf("%s, Your final points are: %d \n", _player.name, _player.points);
+
+        updateLocalFile(&_player, nameOfTheLocalFile);
```
Don't do that, you're manipulating the file every time a player finishes their games. This is to expensive, and to update the file needs more complex code.

hozlucas28 commented on src/play/play.c:
```diff
@@ -0,0 +1,91 @@
+
+        updateLocalFile(&_player, nameOfTheLocalFile);
+
+        SendInformationToTheAPI(&_player);
```
Think about the API expenses, Do you want to do a POST to the API endpoint each every time a player finishes their games? That's a bunch of money, and too slow as it must establish a connection each time.

hozlucas28 commented on src/play/play.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef PLAY_H_INCLUDED
+#define PLAY_H_INCLUDED
```
Invalid header tag, it must be `SRC__PLAY_H_INCLUDED` to prevent tag collisions.

hozlucas28 commented on src/structs.h:
```diff
@@ -19,6 +19,7 @@ typedef struct {
     size_t points;
     char assignedForm;
     size_t remainingTurns;
+    char Turn;  // it will be easer to use a char to represent the turn and having this here
```
Unnecessary, `assignedForm` fulfills that role.

hozlucas28 commented:
- Avoid uploading empty files, they don't add value.
- Folder names must follow [Kebab Case](https://developer.mozilla.org/en-US/docs/Glossary/Kebab_case), its works better with editors and VCS.

ValentinMassa committed:
NEW CHANGES:
- I DELETED THE EMPTY FILE
- I COULDNT FIND THE FILE THAT IS NOT FOLLOWING THE KEBAB CASES PROTOCOL. PLS SHOW TO ME WHERE IS IT.
- I EDITED SOME TEXT

ValentinMassa committed:
I VE DONE A PATTERN FOT THE ISSUES.
LOCATION: .GITHUB/ISSUE_TEMPLATE

ValentinMassa committed:
Some changed about Lucas's comments

ValentinMassa committed:
feature: create play module

ValentinMassa commented:
@jlz10 @parradura @hozlucas28 
Pls confirm pull request changes so this branch can be merged with the master!
